### ISSUE TITLE
B21 not recognized

### ISSUE THREAD

**Issue by spdustin** [MultiMote/niimblue#36]
The [Niimbot B21](https://niimbots.com/products/b21-bluetooth-thermal-label-printer?variant=43580259041516) isn't recognized. After WebBluetooth pairs, niimblue returns the error: `NotFoundError: No Services matching UUID e7810a71-73ae-499d-8c15-faa9aef0c3f2 found in Device.`

**Comment by MultiMote** [comment]
Hi! This is the first issue where Service e7810a71-xxx was not found.
Can you run app like [nRF Connect](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp) and find out what services and characteristics available for your printer?

![ble](https://github.com/user-attachments/assets/cde57716-54c4-49b3-8250-e85795bb60b2)

**Comment by Elcontino** [comment]
Please excuse me but I don't understand what I have to do. I have installed nRF and now?

**Comment by MultiMote** [comment]
@Elcontino I have added UUID to the filter, please try to connect here: https://dev.mmote.ru/niimblue-dev/b21/ and report the result.

**Comment by Elcontino** [comment]
I have done that. Now the label comes out and this error and connection breaks off.
![Screenshot_20250114-165953](https://github.com/user-attachments/assets/f4d49659-7d21-4990-812d-eaa84c42c975)

**Comment by MultiMote** [comment]
Please send a screenshot of gear menu near the printer's name

![image](https://github.com/user-attachments/assets/127444b0-8916-4490-868d-4f590df084f1)

**Comment by Elcontino** [comment]
I dont have this Button 
![Screenshot_20250114-173400](https://github.com/user-attachments/assets/4cd6b65f-bf70-4636-b2c1-990959d13a54)

**Comment by MultiMote** [comment]
Connect to the printer first.

**Comment by Elcontino** [comment]
No longer works 😩

![Uploading Screenshot_20250114-174219.png…]()

**Comment by MultiMote** [comment]
Screenshot link is broken.

**Comment by Elcontino** [comment]
![Screenshot_20250114-174219](https://github.com/user-attachments/assets/00c674ce-5339-42e6-8988-73c5903577ed)

**Comment by MultiMote** [comment]
Updated UUID list. If printer will not connect again, there is no way to make B21 work. Browser implementation requires static UUID list, seems like your printer have dynamic ones.

**Comment by Elcontino** [comment]
Ok where can i Update uuid?

**Comment by MultiMote** [comment]
I did it in the library. Just refresh the page.

**Comment by Elcontino** [comment]
Thanks for your effort, but unfortunately it doesn't work.

**Comment by MultiMote** [comment]
You can download android app from releases, it does not have UUID filter

https://github.com/MultiMote/niimblue/releases/tag/snapshot

**Comment by Elcontino** [comment]
Ok with App i See the Button 
![Uploading Screenshot_20250114-183526.png…]()

**Comment by MultiMote** [comment]
Nice, I will try to find B21_C2B protocol information.
It would help a lot and I will implement it quickly if you can do a dump of a successful print with the NIIMBOT official app (usbpcap software for Windows or BLE Snoop logging for Android).

**Comment by Elcontino** [comment]
Ok, I will try it tomorrow. I didn't manage to connect to the computer today. Thank you very much

**Comment by Elcontino** [comment]
have activated Bluetooth snoop under the developer options. As I have seen, the protocol should be found under etc/Bluetooth. Unfortunately, I cannot find this folder. Any idea what the error is?

**Comment by Elcontino** [comment]
Unfortunately, I cannot establish a connection on my Windows PC. I took an old Android cell phone with root and did everything with it. Unfortunately, I can't find the file. Thanks for your support but after trying for 3 hours now, I have to give up

**Comment by MultiMote** [comment]
If your printer supported by Windows NIIMBOT app, you can capture it's traffic with [usbpcap](https://desowin.org/usbpcap/).
That's much easier, you can just run `USBPcapCMD.exe`, enter hub number (you should find your printer in the list, that's 1 in my case) and enter filename.

![image](https://github.com/user-attachments/assets/d65d34ee-7414-49b9-a40e-5b0972aa9e08)

**Comment by Elcontino** [comment]
Ok thank you. I Try it tomorrow.

**Comment by Elcontino** [comment]
I have seen that the USBPcapCMD.exe no longer exists. I have downloaded the USBPcapSetup-1.5.4.0.exe, which apparently only works with Wireshark. I'am wrong?

**Comment by MultiMote** [comment]
Downloaded the same installer

![image](https://github.com/user-attachments/assets/e78e0481-9ac6-40a0-8b4b-9ee70fff9e7c)

**Comment by Elcontino** [comment]
Yes, but when I start it, it leads me through the setup and then the program cannot be found.

**Comment by MultiMote** [comment]
Did you followed the path on screenshot?

**Comment by Elcontino** [comment]
Ok, I've made some progress but now I'm stuck
![Uploading PXL_20250115_224851499.MP~2.jpg…]()

**Comment by MultiMote** [comment]
It't fine. After entering file name, program starts capturing. Stop it with ctrl+c after you make some print.

**Comment by Elcontino** [comment]
Ok, that's what I did. But nothing is recorded when printing

**Comment by MultiMote** [comment]
Has the file "Niimbot Serial Port" been created (according to your screenshot)? 
It should be in `C:\Program Files\USBPcap` directory if you launched `USBPcapCMD` by double-clicking.

**Comment by Elcontino** [comment]
Yes 
![Uploading 17370264043808108779903488353804.jpg…]()

**Comment by Elcontino** [comment]
Yes 
![Uploading 1000190707.jpg…]()

**Comment by MultiMote** [comment]
This file is what I need. Just attach it if usbpcap was running while printing (rename it to txt in case GitHub blocks upload).

**Comment by Elcontino** [comment]
[niimbot.txt](https://github.com/user-attachments/files/18438763/niimbot.txt)
[Niimbot Serial Port.txt](https://github.com/user-attachments/files/18438764/Niimbot.Serial.Port.txt)

**Comment by MultiMote** [comment]
Perfect, thank you! 
I will analyze it when I get home.

**Comment by Elcontino** [comment]
I have to thank you for your time and patience

**Comment by MultiMote** [comment]
Also, can you open https://niim.mmote.ru, turn the printer on, connect printer to usb, connect to the printer with "Serial" mode (not bluetooth), try to print something, press F12 and copy "Console" tab contents? It will helps with understanding the problem too.

**Comment by Elcontino** [comment]
get the message: failed to open serial Port. I have tried all ports. Does not work with any

**Comment by MultiMote** [comment]
@Elcontino please test this build, is something changed? Delete installed app first.

[app-release-signed.zip](https://github.com/user-attachments/files/18481740/app-release-signed.zip)

How the printer behaves after you press Print?

**Comment by Elcontino** [comment]
So the paper comes out when printing, but the paper itself is not printed

**Comment by MultiMote** [comment]
Same for previous version or something changed?

**Comment by Elcontino** [comment]
The difference is that now no error message appears and the printer remains connected

**Comment by Elcontino** [comment]
But no Text on the label.

**Comment by MultiMote** [comment]
Try to set 48x30mm label size. I don't see any major differences between the niimblue protocol and the official application protocol.

**Comment by Elcontino** [comment]
Error: Print error 6: DataError 😩

**Comment by MultiMote** [comment]
Noticed that B21 doesn't use some packet in the image encoding. Deleted it, it may be causing the error. Also added some debug details to the error message.

[app-release-signed.zip](https://github.com/user-attachments/files/18494099/app-release-signed.zip)

**Comment by Elcontino** [comment]
I Try it and i receive this error: Error: Print error 6: DataError, last sent packet is 5555e30101e3aaaa.

After that, i set 48x30 and the niimbot print but without Text on the label.

**Comment by MultiMote** [comment]
Another attempt, implemented periodical image line check that is possibly needed for your printer

[app-release-signed.zip](https://github.com/user-attachments/files/18495956/app-release-signed.zip)

**Comment by Elcontino** [comment]
Ok thanks

**Comment by MultiMote** [comment]
Let me know if anything changed.

**Comment by thewan056** [comment]
> Another attempt, implemented periodical image row checking that is possibly needed for your printer
> 
> [app-release-signed.zip](https://github.com/user-attachments/files/18495956/app-release-signed.zip)

Hi. I have a B21 and I tried this version. It works fine for me. The label printed with both image and text.

Is it possible to get the web version to work? I tried the web version first with both bluetooth and serial it didnt work. Searching issues brought me here.

![Image](https://github.com/user-attachments/assets/0d6656df-792d-4bf3-865d-0015864d8d64)

**Comment by MultiMote** [comment]
@thewan056 Hi, can you provide more details? What kind of error do you get and in what case?

Also attach browser console logs (F12, console tab).

I think it's better to create a new Issue.
